code/drasil-database/lib/Database/Drasil/ChunkDB.hs

```diff
@@ -1,312 +1,274 @@
--- what the map is for. One is for symbols + their units, and the others are for
--- what they state.
-type UMap a = Map.Map UID (a, Int)
-
--- | A bit of a misnomer as it's really a map of all quantities, for retrieving
--- symbols and their units.
-type SymbolMap  = UMap DefinedQuantityDict
-
--- | A map of all concepts, normally used for retrieving definitions.
-type ConceptMap = UMap ConceptChunk
-
--- | A map of all the units used. Should be restricted to base units/synonyms.
-type UnitMap = UMap UnitDefn
-
--- | Again a bit of a misnomer as it's really a map of all 'NamedIdea's.
--- Until these are built through automated means, there will
--- likely be some 'manual' duplication of terms as this map will contain all
--- quantities, concepts, etc.
-type TermMap = UMap IdeaDict
--- | A traceability map, used to hold the relation between one 'UID' and a list of other 'UID's.
-type TraceMap = Map.Map UID [UID]
--- | A reference map, used to hold a 'UID' and where it is referenced ('UID's).
-type RefbyMap = Map.Map UID [UID]
--- | Data definitions map. Contains all data definitions ('DataDefinition').
-type DatadefnMap = UMap DataDefinition
--- | Instance model map. Contains all instance models ('InstanceModel').
-type InsModelMap = UMap InstanceModel
--- | General definitions map. Contains all general definitions ('GenDefn').
-type GendefMap = UMap GenDefn
--- | Theory model map. Contains all theoretical models ('TheoryModel').
-type TheoryModelMap = UMap TheoryModel
--- | Concept instance map. May hold similar information to a 'ConceptMap', but may also be referred to.
-type ConceptInstanceMap = UMap ConceptInstance
--- | A map of all 'LabelledContent's.
-type LabelledContentMap = UMap LabelledContent
--- | A map of all 'Reference's.
-type ReferenceMap = UMap Reference
--- | Citation map.
-type CitationMap = UMap Citation
-
--- | General chunk database map constructor. Creates a 'UMap' from a function
--- that converts something with 'UID's into another type and a list of something
--- with 'UID's.
-cdbMap :: HasUID a => String -> (a -> b) -> [a] -> Map.Map UID (b, Int)
-cdbMap mn f rawChunks = Map.fromListWithKey preferNew kvs
+-- NOTE: Strictness is important for (a) performance, (b) space leaks, (c)
+-- avoiding chunk dependancy cycles and (d) ensuring operation consistency with
+-- other databases.
+import qualified Data.Map.Strict as M
+
+import Drasil.Database.Chunk (Chunk, HasChunkRefs(chunkRefs), IsChunk,
+  mkChunk, unChunk, chunkType)
+import Language.Drasil (HasUID(..), UID, LabelledContent, Reference)
+import Utils.Drasil (errMsg, warnMsg)
+
+-- | A chunk that depends on another.
+type Dependant = UID
+
+-- | Mapping of 'UID's to 'Chunk's and their dependants.
+type ChunkByUID = M.Map UID (Chunk, [Dependant]) -- FIXME: the list of dependants should be a Set.
+
+-- | Mapping of chunk types to lists of instances of them (chunks).
+type ChunksByTypeRep = M.Map TypeRep [Chunk]
```

Comment:
TODO: @balacij Instead of lists here, use sets.

Reply:
Actually, maybe it's not quite worth it. I'll avoid preemptive optimizing. We were fine with lists before, we will be fine with lists going for a while longer.